hypothesis-python/src/hypothesis/internal/reflection.py

```diff
@@ -281,32 +285,60 @@ def is_first_param_referenced_in_function(f: Any) -> bool:
     )
 
 
-def extract_all_lambdas(tree, matching_signature):
+def extract_all_lambdas(tree):
     lambdas = []
 
     class Visitor(ast.NodeVisitor):
         def visit_Lambda(self, node):
-            if ast_arguments_matches_signature(node.args, matching_signature):
-                lambdas.append(node)
+            lambdas.append(node)
 
     Visitor().visit(tree)
-
     return lambdas
 
 
-LINE_CONTINUATION = re.compile(r"\\\n")
-WHITESPACE = re.compile(r"\s+")
-PROBABLY_A_COMMENT = re.compile("""#[^'"]*$""")
-SPACE_FOLLOWS_OPEN_BRACKET = re.compile(r"\( ")
-SPACE_PRECEDES_CLOSE_BRACKET = re.compile(r" \)")
+def _lambda_source_key(f, *, bounded_size=False):
+    """Returns a digest that differentiates lambdas that have different sources."""
+    consts_repr = repr(f.__code__.co_consts)
+    if bounded_size:
+        consts_repr = hashlib.sha384(consts_repr.encode()).digest()
+    return (
+        consts_repr,
+        inspect.signature(f),
+        f.__code__.co_names,
+        f.__code__.co_code,
+        f.__code__.co_varnames,
+        f.__code__.co_freevars,
+    )
+
+
+def _mimic_lambda_from_source(f, source):
+    # Compile a lambda from source where the compiled lambda mimics f as far as
+    # possible in terms of __code__ and __closure__. The mimicry is far from perfect.
+    if f.__closure__ is None:
+        compiled = eval(source, f.__globals__)
+    else:
+        # Hack to mimic the capture of vars from local closure. In terms of code
+        # generation they don't *need* to have the same values (cell_contents),
+        # just the same names bound to some closure, but hey.
+        closure = {f"___fv{i}": c.cell_contents for i, c in enumerate(f.__closure__)}
+        assigns = [f"{name}=___fv{i}" for i, name in enumerate(f.__code__.co_freevars)]
+        fake_globals = f.__globals__ | closure
+        exec(f"def construct(): {';'.join(assigns)}; return ({source})", fake_globals)
+        compiled = fake_globals["construct"]()
+    return compiled
+
+
+def _lambda_code_matches_source(f, source):
+    try:
+        compiled = _mimic_lambda_from_source(f, source)
+    except (NameError, SyntaxError):  # pragma: no cover
+        return False
+    return _lambda_source_key(f) == _lambda_source_key(compiled)
 
 
 def _extract_lambda_source(f):
     """Extracts a single lambda expression from the string source. Returns a
     string indicating an unknown body if it gets confused in any way.
-
-    This is not a good function and I am sorry for it. Forgive me my
-    sins, oh lord
```

Comment:
I took the liberty of removing this humble comment, but I might just move it to `_mimic_lambda_from_source` instead :-)